### ISSUE TITLE
Impossible to set a static IPv6 address on an interface

### ISSUE THREAD

**Issue by eriksl** [espressif/esp-idf#14012]
### Answers checklist.

- [X] I have read the documentation [ESP-IDF Programming Guide](https://docs.espressif.com/projects/esp-idf/en/latest/) and the issue is not addressed there.
- [X] I have updated my IDF branch (master or release) to the latest version and checked that the issue is present there.
- [X] I have searched the issue tracker for a similar issue and not found a similar issue.

### General issue report

This is related to "Static ip address of IPV6 (IDFGH-11369) #12515".

With the comment there, it's probably possible to set a static IPv6 address. The definition of this function still needs to be moved to a public header so we can actually use it, and it will be included in the documentation.

But besides that, there is no way to add a netmask or a gateway to the IPv6 address. I assume that LWIP assumes netmask for ipv6 is always 64 bits. This is an invalid assumption but I guess Espressif can't fix that (other than put some pressure on the LWIP guys).

Also there is no way to specify the default gateway. Using SLAAC this works, but I don't want to use SLAAC because it can only work on /64 netmasks AND I don't like the long addresses (I prefer using face:b00c:1::1 style).

It might work if you combine SLAAC to obtain the default gw and then add an additional, */64 netmasked* address, is that correct? So it can never work without SLAAC...

**Comment by eriksl** [comment]
Update, it appears to work without SLAAC (a static address). But only when the netmask is /64. I am wondering how the default gateway is determined. Peeking at incoming non-local packets? Sending to ff02::2 and see who replies? Kind of ugly in my opinion.

**Comment by abhik-roy85** [comment]
Hi @eriksl,

 Apologies for the delayed response!

`esp_netif_add_ip6_address()` is already a public, documented API see [ESP-NETIF Programmers Manual](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/network/esp_netif.html). It lets you add a static global unicast IPv6 address and mark it as preferred.
However, there are some important limitations inherited from lwIP:
* The stack assumes /64 for unicast IPv6 addresses when added this way. There is no parameter in the API (or in lwIP's equivalent netif_add_ip6_address) to specify a different prefix length. Non-/64 prefixes are not supported for global unicast addresses in this context. (Link-local addresses are always /64 by definition.)
* There is currently no esp-netif API to set a static IPv6 default gateway or static routes. Gateway information normally comes from Router Advertisements (RA) via SLAAC/ND.
Without RA, any default route is typically added heuristically (e.g. from observed packets), which is unreliable and not a proper static solution.

We'll improve the documentation to clearly cover these limitations (/64 requirement for static global addresses, lack of static gateway API, and RA dependency).

**Comment by eriksl** [comment]
Yes, that is what I figured out myself in the meantime. Nice to update the documentation with these facts.

I am used to work in a networking environment (switches and routers) where it's "quite common" to configure (ipv6) addresses, gateways and netmasks. For a client, indeed, this is not strictly required, especially when you're sending out RA's. This is what I already did, so no problem there. I have adjusted the netmask/prefix to 64 as well, simply because a lot of devices assume 64 bit, even though there is no actual technical reason to enforce a 64 bit netmask (unless you're using SLAAC, which I am not using).

So my approach is now like this:

- RA active on the vlan
- 64 bit prefix
- disable SLAAC
- add static address and inherit gw/prefix from RA

This works, but there is a flaw, which, IIRC, I mentioned in another ticket. Disabling SLAAC **does not work**. I don't mind the SLAAC (although I think it's ugly) but it becomes the **preferred** address, no matter what I do. When I add a static address and mark it as "preferred", the SLAAC address still remains the preferred address. This means that if I send a UDP packet to it's static address, the response will be sent using the SLAAC address. That's a source for all kinds of problems.

So for now I've worked around this by obtaining the LWIP handle and setting "autodiscovery" forced and THEN it works without a problem.

I guess Espressif can fix this issue via several approaches, one of them being fixing the SLAAC address always becoming the preferred address. Or fixing I cannot turn of SLAAC via the menuconfig (to be clear: yes you can disable the item, but it doesn't make a difference).

**Comment by eriksl** [comment]
Will anybody have a look into the issues I described in the last post?
